commit a22a7c707765fdc83bfdbe890e01b45b8123ab40 (refs/changes/78/3278/4)
Author: Kody A Kantor <kody.kantor@gmail.com>
Date:   2018-01-30T20:01:50+00:00 (1 year, 8 months ago)
    
    joyent/node-artedi#12 histograms can't observe values less than one

diff --git a/CHANGES.md b/CHANGES.md
index b115030..48d9fba 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -3,6 +3,9 @@
 ## Not yet released
 None
 
+## 1.2.1
+* #12 histograms can't observe values less than one
+
 ## 1.2.0
 * #9 Create an accessor function for collector values
 
diff --git a/README.md b/README.md
index daf0c3b..0ca283b 100644
--- a/README.md
+++ b/README.md
@@ -20,7 +20,7 @@ var counter = collector.counter({
     name: 'http_requests_completed',
     help: 'count of muskie http requests completed',
     labels: {
-        zone: ZONENAME
+        zone: 'e5d3'
     }
 });
 
@@ -53,7 +53,7 @@ histogram.observe(998, {
 // This output is defined by Prometheus.
 collector.collect(artedi.FMT_PROM, function (err, metrics) {
     if (err) {
-        throw new VError(err, 'could not collect metrics');
+        throw new Error('could not collect metrics');
     }
     console.log(metrics);
     // Prints:
@@ -64,9 +64,9 @@ collector.collect(artedi.FMT_PROM, function (err, metrics) {
     // # TYPE http_request_latency_ms histogram
     // http_request_latency_ms{zone="e5d3",method="getobject",code="200",le="729"} 0
     // http_request_latency_ms{zone="e5d3",method="getobject",code="200",le="2187"} 1
-    // http_request_latency_ms{zone="e5d3",method="getobject",code="200",le="3645"} 0
-    // http_request_latency_ms{zone="e5d3",method="getobject",code="200",le="5103"} 0
-    // http_request_latency_ms{zone="e5d3",method="getobject",code="200",le="6561"} 0
+    // http_request_latency_ms{zone="e5d3",method="getobject",code="200",le="3645"} 1
+    // http_request_latency_ms{zone="e5d3",method="getobject",code="200",le="5103"} 1
+    // http_request_latency_ms{zone="e5d3",method="getobject",code="200",le="6561"} 1
     // http_request_latency_ms{zone="e5d3",method="getobject",code="200",le="+Inf"} 1
     // http_request_latency_ms_count{zone="e5d3",method="getobject",code="200"} 1
     // http_request_latency_ms_sum{zone="e5d3",method="getobject",code="200"} 998
diff --git a/lib/histogram.js b/lib/histogram.js
index f8ef86d..7fa98b9 100644
--- a/lib/histogram.js
+++ b/lib/histogram.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 var mod_util = require('util');
 
@@ -286,10 +286,13 @@ Histogram.prototype.defaultCounter = function defaultCounter() {
  * The 'steps' argument is provided, as we may allow the user to specify
  * the number of linear 'steps' between logarithmic jumps. This idea is taken
  * from DTrace's log/linear quantization ('llquantize()').
+ *
+ * The 'steps' argument is ignored for retrieving the order of values less than
+ * one.
  */
 function getOrder(value, steps) {
-    var i, j, val, width, next;
-    var buckets = [];
+    var i, j, bucketVal, width, next;
+    var buckets;
 
     // These values are relatively arbitrary. It's possible that we may want to
     // allow the user to change these at some point.
@@ -299,21 +302,35 @@ function getOrder(value, steps) {
     // This is an arbitrary high watermark. Setting this allows us to ensure
     // that our loop will always exit.
     // The maximum value that we can observe (without it falling into only the
-    // +Inf bucket is 10^10, or 10,000,000,000.
-    var high = 10;
+    // +Inf bucket) is 3,486,784,401
+    var high = 13;
 
-    val = 1;
+    bucketVal = 0.0001; // smallest possible precision.
     for (i = low; i <= high; i++) {
-        next = val * factor;
-        width = next > steps ? next / steps : 1;
+        buckets = [];
+
+        next = bucketVal * factor;
+        if (bucketVal < 1) {
+            // Ignore the 'step' count for very small values.
+            width = bucketVal;
+        } else {
+            // Otherwise enforce that we don't have more than 'step' buckets.
+            width = next > steps ? next / steps : 1;
+        }
 
-        for (j = 0; val <= next; val += width, j++) {
-            buckets[j] = val;
+        for (j = 0; bucketVal <= next; bucketVal += width, j++) {
+            if (bucketVal < 10) {
+                // only keep a few decimal places when decimal precision matters
+                bucketVal = +(bucketVal.toFixed(4));
+            } else {
+                bucketVal = Math.ceil(bucketVal);
+            }
+            buckets[j] = bucketVal;
         }
 
         // Overlap buckets so we get something like:
         // [0-10], [10-100], [100-1000].
-        val -= width;
+        bucketVal -= width;
         if (value <= buckets[buckets.length - 1]) {
             // The number is within this order.
             return (buckets);
diff --git a/package.json b/package.json
index 41bdd77..1411eda 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "artedi",
-  "version": "1.2.0",
+  "version": "1.2.1",
   "description": "a metric client library",
   "main": "lib/collector.js",
   "dependencies": {
diff --git a/test/basic.test.js b/test/basic.test.js
index c08f297..5455121 100644
--- a/test/basic.test.js
+++ b/test/basic.test.js
@@ -357,14 +357,19 @@ mod_tape('histogram serialization tests', function (t) {
             });
 
             expected = expected +
-                'bot_trolololol{key="value",le="1"} 1\n' +
-                'bot_trolololol{key="value",le="3"} 1\n' +
-                'bot_trolololol{key="value",le="5"} 1\n' +
-                'bot_trolololol{key="value",le="7"} 1\n' +
-                'bot_trolololol{key="value",le="9"} 1\n' +
-                'bot_trolololol{le="+Inf",key="value"} 1\n' +
-                'bot_trolololol_count{key="value"} 1\n' +
-                'bot_trolololol_sum{key="value"} 1\n';
+                'bot_trolololol{key="value",le="0.81"} 0\n'
+                + 'bot_trolololol{key="value",le="1.62"} 1\n'
+                + 'bot_trolololol{key="value",le="2.43"} 1\n'
+                + 'bot_trolololol{key="value",le="3.24"} 1\n'
+                + 'bot_trolololol{key="value",le="4.05"} 1\n'
+                + 'bot_trolololol{key="value",le="4.86"} 1\n'
+                + 'bot_trolololol{key="value",le="5.67"} 1\n'
+                + 'bot_trolololol{key="value",le="6.48"} 1\n'
+                + 'bot_trolololol{key="value",le="7.29"} 1\n'
+                + 'bot_trolololol{key="value",le="8.1"} 1\n'
+                + 'bot_trolololol{le="+Inf",key="value"} 1\n'
+                + 'bot_trolololol_count{key="value"} 1\n'
+                + 'bot_trolololol_sum{key="value"} 1\n';
 
             collector.collect(mod_artedi.FMT_PROM, function (err, str) {
                 t.notOk(err, 'no error for single label');
@@ -390,11 +395,11 @@ mod_tape('histogram serialization tests', function (t) {
             expected = '' +
                 '# HELP http_request_latency latency of requests\n' +
                 '# TYPE http_request_latency ' + common.HISTOGRAM + '\n' +
-                'http_request_latency{service="muskie",le="81"} 0\n' +
-                'http_request_latency{service="muskie",le="243"} 1\n' +
-                'http_request_latency{service="muskie",le="405"} 1\n' +
-                'http_request_latency{service="muskie",le="567"} 1\n' +
-                'http_request_latency{service="muskie",le="729"} 1\n' +
+                'http_request_latency{service="muskie",le="76"} 0\n' +
+                'http_request_latency{service="muskie",le="228"} 1\n' +
+                'http_request_latency{service="muskie",le="380"} 1\n' +
+                'http_request_latency{service="muskie",le="532"} 1\n' +
+                'http_request_latency{service="muskie",le="684"} 1\n' +
                 'http_request_latency{le="+Inf",service="muskie"} 1\n' +
                 'http_request_latency_count{service="muskie"} 1\n' +
                 'http_request_latency_sum{service="muskie"} 99\n';
@@ -430,11 +435,11 @@ mod_tape('histogram serialization tests', function (t) {
             expected = '' +
             '# HELP web_conn_alive_time connection alive time\n' +
             '# TYPE web_conn_alive_time ' + common.HISTOGRAM + '\n' +
-            'web_conn_alive_time{service="muskie",component="qball",le="81"} 0\n' +
-            'web_conn_alive_time{service="muskie",component="qball",le="243"} 1\n' +
-            'web_conn_alive_time{service="muskie",component="qball",le="405"} 1\n' +
-            'web_conn_alive_time{service="muskie",component="qball",le="567"} 1\n' +
-            'web_conn_alive_time{service="muskie",component="qball",le="729"} 1\n' +
+            'web_conn_alive_time{service="muskie",component="qball",le="76"} 0\n' +
+            'web_conn_alive_time{service="muskie",component="qball",le="228"} 1\n' +
+            'web_conn_alive_time{service="muskie",component="qball",le="380"} 1\n' +
+            'web_conn_alive_time{service="muskie",component="qball",le="532"} 1\n' +
+            'web_conn_alive_time{service="muskie",component="qball",le="684"} 1\n' +
             'web_conn_alive_time{le="+Inf",service="muskie",component="qball"} 1\n' +
             'web_conn_alive_time_count{service="muskie",component="qball"} 1\n' +
             'web_conn_alive_time_sum{service="muskie",component="qball"} 101\n';
@@ -476,11 +481,11 @@ mod_tape('histogram serialization tests', function (t) {
             var expected4 = '' +
             '# HELP webapi_conn_alive_time connection alive time\n' +
             '# TYPE webapi_conn_alive_time ' + common.HISTOGRAM + '\n' +
-            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="81"} 0\n' +
-            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="243"} 1\n' +
-            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="405"} 1\n' +
-            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="567"} 1\n' +
-            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="729"} 1\n' +
+            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="76"} 0\n' +
+            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="228"} 1\n' +
+            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="380"} 1\n' +
+            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="532"} 1\n' +
+            'webapi_conn_alive_time{err="ECONNRESET",service="muskie",component="cueball",le="684"} 1\n' +
             'webapi_conn_alive_time{le="+Inf",err="ECONNRESET",service="muskie",component="cueball"} 1\n' +
             'webapi_conn_alive_time_count{err="ECONNRESET",service="muskie",component="cueball"} 1\n' +
             'webapi_conn_alive_time_sum{err="ECONNRESET",service="muskie",component="cueball"} 101\n';
@@ -502,27 +507,27 @@ mod_tape('histogram serialization tests', function (t) {
                 help: 'testhelp'
             });
 
-            histogram.observe(1);
+            histogram.observe(10);
             histogram.observe(100);
             // TODO We should have the +Inf label at the end. This works, but
             // it would look nicer.
             expected = '' +
-                '# HELP test_test1 testhelp\n' +
-                '# TYPE test_test1 ' + common.HISTOGRAM + '\n' +
-                'test_test1{le="1"} 1\n' +
-                'test_test1{le="3"} 1\n' +
-                'test_test1{le="5"} 1\n' +
-                'test_test1{le="7"} 1\n' +
-                'test_test1{le="9"} 1\n' +
-                'test_test1{le="+Inf"} 2\n' +
-                'test_test1{le="81"} 1\n' +
-                'test_test1{le="243"} 2\n' +
-                'test_test1{le="405"} 2\n' +
-                'test_test1{le="567"} 2\n' +
-                'test_test1{le="729"} 2\n' +
-                'test_test1_count{} 2\n' +
-                'test_test1_sum{} 101\n';
-            collector.collect(mod_artedi.FMT_PROM, function (err, str) {
+                '# HELP test_test1 testhelp\n'
+                + '# TYPE test_test1 ' + common.HISTOGRAM + '\n'
+                + 'test_test1{le="8.1"} 0\n'
+                + 'test_test1{le="25"} 1\n'
+                + 'test_test1{le="42"} 1\n'
+                + 'test_test1{le="59"} 1\n'
+                + 'test_test1{le="76"} 1\n'
+                + 'test_test1{le="+Inf"} 2\n'
+                + 'test_test1{le="228"} 2\n'
+                + 'test_test1{le="380"} 2\n'
+                + 'test_test1{le="532"} 2\n'
+                + 'test_test1{le="684"} 2\n'
+                + 'test_test1_count{} 2\n'
+                + 'test_test1_sum{} 110\n';
+
+                collector.collect(mod_artedi.FMT_PROM, function (err, str) {
                 t.notOk(err, 'no error for copying bucket values');
                 t.equals(str, expected, 'initial values copied from ' +
                     'low-order buckets to high-order buckets');
@@ -549,6 +554,10 @@ mod_tape('odd value tests', function (t) {
     counter.add(0);
     t.equals(counter.getValue(), 0, 'add zero to counter');
 
+    hist.observe(0.0001);
+    t.equals(hist.defaultCounter().getValue({'le': 0.0001}), 1,
+        'histogram tracks values less than one');
+
     hist.observe(0);
     t.equals(hist.defaultCounter().getValue(), 0,
         'histogram observes zero value');
@@ -581,6 +590,7 @@ mod_tape('odd value tests', function (t) {
         });
     }, 'invalid collector help (object)');
 
+
     t.end();
 });
 
@@ -646,10 +656,10 @@ mod_tape('log/linear bucket tests', function (t) {
     histo.observe(10); // Record three small values.
     histo.observe(10);
     histo.observe(10);
-    histo.observe(6562); // Record a value which will create larger buckets.
-    histo.observe(5103); // Record a value below the smallest larger bucket.
+    histo.observe(6157); // Record a value which will create larger buckets.
+    histo.observe(4788); // Record a value below the smallest larger bucket.
 
-    value = histo.defaultCounter().getValue({'le': 6561});
+    value = histo.defaultCounter().getValue({'le': 6156});
     t.equals(value, 4, 'overlapping bucket values copied correctly');
 
 
